dataset/src/test/scala/frameless/ops/SmartProjectTest.scala

```diff
@@ -0,0 +1,48 @@
+package frameless
+package ops
+
+import org.scalacheck.Prop
+import org.scalacheck.Prop._
+import shapeless.test.illTyped
+
+
+case class Foo(i: Int, j: Int, x: String)
+case class Bar(i: Int, x: String)
+case class InvalidFooProjectionType(i: Int, x: Boolean)
+case class InvalidFooProjectionName(i: Int, xerr: String)
+
+class SmartProjectTest extends TypedDatasetSuite {
+  val dataset = TypedDataset.create(Foo(1, 2, "hi") :: Foo(2, 3, "there") :: Nil)
+
+  test("project Foo to Bar") {
+    assert(dataset.project[Bar].count().run() === 2)
+  }
+
+  test("project to InvalidFooProjection should not type check") {
+    illTyped("dataset.project[InvalidFooProjectionType]")
+    illTyped("dataset.project[InvalidFooProjectionName]")
+  }
+
+  test("X4 to X1,X2,X3,X4 projections") {
+    def prop[A, B, C, D](data: Vector[X4[A, B, C, D]])(
```

Comment:
I guess instead of requiring `ex4: TypedEncoder[X4[A, B, C, D]],` and etc, it should be enough to require `TypedEncoder` for for `A`, `B`, `C`, `D`:
```scala
def prop[A: TypedEncoder, B: TypedEncoder, C: TypedEncoder, D: TypedEncoder](data: Vector[X4[A, B, C, D]])(
```